doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,906 @@
+
+## Requirements / Constraints
+
+- The number of SAI object types should not exceed 32,768 ($2^{15}$). This means the value of SAI_OBJECT_TYPE_MAX should be less than 32,768.
+- The number of SAI object extension types should not exceed 32,768.
+- The number of stats types for a single SAI object type should not exceed 32,768.
+- The number of extension stats types for a single SAI object type should not exceed 32,768.
+- The number of SAI objects of the same type should not exceed 32,768.
+- The vendor SDK should support publishing stats in IPFIX format and its IPFIX template.
+- If a polling frequency for stats cannot be supported, the vendor's SDK should return this error.
+- The vendor SDK should support querying the minimal polling interval for each counter.
+- When reconfiguring any stream settings, whether it is the polling interval or the stats list, the existing stream will be interrupted and regenerated.
+
+## Architecture Design
+
+``` mermaid
+
+---
+title: Stream telemetry architecture
+---
+flowchart BT
+    subgraph Userspace
+        rec(recording collect process)
+        cfg(SAI config process)
+    end
+
+    subgraph Linux Kernel
+        netlink_module(Netlink Module)
+    end
+
+    asic[\ASIC\]
+
+    cfg --IPFIX template--> rec
+    cfg --TAM configuration--> netlink_module
+    asic --counters--> netlink_module
+    netlink_module --IPFIX record--> rec
+```
+
+## High-Level Design
+
+### Modules
+
+#### recording collect process
+
+The primary responsibility of `CounterSyncd`,a.k.a `recording collect process`, is to receive counter messages via netlink and convert them into telemetry messages for an external collector. It subscribes to a socket of a specific family and multicast group of generic netlink.
+
+#### Netlink Module
+
+These module needs to be provided by vendors. This document proposes a ring buffer communication model to support all expected TAM configurations as follows.
+
+![netlink_dma_channel](netlink_dma_channel.drawio.svg)
+
+### Data format
+
+We will use IPFIX as the report format, with all numbers in the IPFIX message in network-order (Big-endian).
+
+For more information on IPFIX, refer to the following resources:
+
+- [Specification of the IP Flow Information Export (IPFIX) Protocol for the Exchange of Flow Information](https://datatracker.ietf.org/doc/html/rfc7011)
+- [IP Flow Information Export (IPFIX) Entities](https://www.iana.org/assignments/ipfix/ipfix.xhtml)
+
+#### IPFIX header
+
+The `Version` and `Observation Domain ID` fields of the IPFIX header are identical for each IPFIX message.
+
+``` mermaid
+
+---
+title: stream message IPFIX header
+---
+packet-beta
+0-15: "Version = 0x000a"
+16-31: "Message Length = (16 + payload) bytes"
+32-63: "Export Timestamp: Second"
+64-95: "Sequence Number = 0, start from 0 and incremental sequence counter modulo 2^32"
+96-127: "Observation Domain ID = 0, always 0"
+
+```
+
+#### IPFIX template

```

Comment:
We only expect that the IPFIX template must include these stat-IDs that we explicitly configured. But if there are more entries included in the template, I feel we can ignore it.

But I feel there are two concerns:
1. what's the label(element ID) will be assigned? Because we pass the label to vendor's SDK via `SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STAT_ID`
2. If it includes too much un-required data, the efficiency may be decreased dramatically.

@r12f Please correct me if you have any suggestions.

Reply:
> 2\. If it includes too much un-required

1. Since label and SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STAT_ID have fixed encoding, SAI implementation can pass the additional info (that are not subscribed) in the same format and these additional records should be ignored.
2. Agree that un-wanted data will cause performance issues.

In general,  this proposal talks about only SAI_TAM_TELEMETRY_TYPE_COUNTER_SUBSCRIPTION. Do you have plans to extend this proposal to support other telemetry types ?

Reply:
What's the meaning about other telemetry types?
We have `sai_tam_report_type_t`, which can support 
```c
// ...
    /**
     * @brief Report using IPFIX
     */
    SAI_TAM_REPORT_TYPE_IPFIX,

    /**
     * @brief Report using GPB
     */
    SAI_TAM_REPORT_TYPE_PROTO,

    /**
     * @brief Report using THRIFT
     */
    SAI_TAM_REPORT_TYPE_THRIFT,

    /**
     * @brief Report using JSON
     */
    SAI_TAM_REPORT_TYPE_JSON,

    /**
     * @brief Report using P4 format
     */
    SAI_TAM_REPORT_TYPE_P4_EXTN,

    /**
     * @brief Report using Histogram
     */
    SAI_TAM_REPORT_TYPE_HISTOGRAM,
// ...
```
So, I believe the current SAI is flexible enough to support other telemetry types. But I don't have a short-term plan to discuss other types in this proposal.

Reply:
By other telemetry type, I meant below. For example, SAI_TAM_TELEMETRY_TYPE_SWITCH can as well fetch all supported statistics from all ports in the switch. But, in this type, LABEL (element_id) cannot be mentioned by NOS. If SAI_TAM_TELEMETRY_TYPE_SWITCH is applicable, can you clarify what should be IPFIX template encoding for element-id, enterprise number ? 

```
    /**
     * @brief Switch silicon TAM
     * All the data relevant to switch
     * e.g. route, port, queue statistics
     */
    SAI_TAM_TELEMETRY_TYPE_SWITCH,

```